### PR TITLE
[PyTorch] Add Support for Embedding Layers

### PR DESCRIPTION
**Pull request by Atrisan**
## Description
Add support for embeddings includes support for the following aten ops:
* embedding.default
* embedding_renorm.default
* alias.default
* detach.default

Adds the following new tensor nodes:
* embedding
* embedding renorm

Only return user outputs, to avoid conflicts with inplace ops like embedding_renorm.default.

## Related Issues

- Closes #

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [x] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [x] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [x]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [x] New / modified types have serialization / deserialization support
  *  [x] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [x] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [x] Serializer has an option to emit the previous format

### Unit Tests
- [x] New unit tests have been added for the changes.
- [x] Existing unit tests pass locally.
- [x] Edge cases and error paths are covered.

### Integration Tests
- [x] New integration tests have been added (or existing ones updated).
- [x] Integration tests pass locally.
- [x] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [x] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [x] Test coverage has not decreased as a result of this change.
- [x] Coverage report reviewed and acceptable.
- [x] Critical / new code paths are covered.

## How Has This Been Tested?
Unit tests for libnodes.
Integration tests for ops.

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: sdfg/src/data_flow/library_nodes/math/tensor/embedding_node.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+    if (weight_shape_.size() != 2) {
+        throw InvalidSDFGException(
+            "EmbeddingNode: weight must be 2-dimensional but got rank " + std::to_string(weight_shape_.size())
+        );
+    }
+    if (index_shape_.empty()) {
+        throw InvalidSDFGException("EmbeddingNode: index_shape must not be empty");
+    }
+}
```
Comment (WashimNeupane) [inline sdfg/src/data_flow/library_nodes/math/tensor/embedding_node.cpp:42]: Maybe add connector type check here since we are skipping TensorNode::validate(). At minimum, it should verify that "I" edge is an integer type and "Y/W" edge are float/non-integer

File: pytorch/docc/pytorch/graph_parser/__init__.py [hunk 2]
```diff
@@ -66,12 +66,21 @@ def get_output_containers(self, node: torch.fx.Node, args: Argument) -> list[str
...
         """
         result = []
         if isinstance(args, tuple):
             for elem in list(args):
                 result += self.get_output_containers(node, elem)
         else:
+            if isinstance(args, torch.fx.Node):
+                user_outputs = getattr(self.ep.graph_signature, "user_outputs", None)
+                if user_outputs is not None and args.name not in user_outputs:
+                    return result
```
Comment (WashimNeupane) [inline pytorch/docc/pytorch/graph_parser/__init__.py:84]: The check is written generically for any output node that isn't in user_outputs. If someone adds another op tomorrow that unintentionally lands outside user_outputs (perhaps due to a bug), this code will quietly discard it instead of raising an error, making the bug very hard to find. maybe we have to raise an error when an unknown non-user-visible output is encountered.

File: sdfg/src/data_flow/library_nodes/math/tensor/embedding_node.cpp (hunk 1 above)
Comment (Atrisan) [reply sdfg/src/data_flow/library_nodes/math/tensor/embedding_node.cpp:42]: I is reasonable, but this tensor node support int as well so Y/W can be int and float.
See line 25:
bool EmbeddingNode::supports_integer_types() const { return true; }

File: pytorch/docc/pytorch/graph_parser/__init__.py (hunk 2 above)
Comment (Atrisan) [reply pytorch/docc/pytorch/graph_parser/__init__.py:84]: This is exactly the idea.
Because embedding_renorm is an inplace operation you run for its side effect.
As a result, its input will be considered an output as well, for the entire model.
Even though it is not a real output.

The problem is that you need to filter those to ensure analogous outputs to plain PyTorch.
I'll have to check if this becomes redundant, when the output reconstruction is merged.
Until then, we need to ensure that we do not add new tupel elements to the output.

I am also not sure if this is the best solution to this problem either.

**Review by Moehre2 — APPROVED** [review]
